Take advantage of known UTF-8 encoding to speed up strings

dmk42 opened the pull request:
This change speeds up the substring operation by more than 3x to address string performance regressions discussed in #13130 .

The String module was originally updated in a character-set-agnostic way.  Now that we know the character set is guaranteed to be UTF-8, we can take advantage of that fact not just to speed up individual operations, but to choose faster algorithms.  For example, there are now some cases where a multibyte character does not have to be decoded, where it was before.

The 3x performance benefit was observed in the substring portion of `test/types/string/psahabu/perf/substring.chpl`.

Passes full local and GASNet testing.

dmk42 commented:
@mppf - Would you mind reviewing this?

@e-kayrakli - FYI.

mppf commented on modules/internal/String.chpl:
```diff
@@ -838,24 +854,52 @@ module String {
       Iterates over the string Unicode character by Unicode character,
       and includes the byte index and byte length of each character.
       Skip characters that begin prior to the specified starting byte index.
+      Assume we accidentally start in the middle of a multibyte character,
```
-> "assume we may" ?

mppf commented on modules/internal/String.chpl:
```diff
@@ -838,24 +854,52 @@ module String {
+      Iterates over the string Unicode character by Unicode character,
+      and returns the byte index and byte length of each character.
+      Skip characters that begin prior to the specified starting byte index.
+      Assume we accidentally start in the middle of a multibyte character,
```
same here

dmk42 commented:
@mppf - Thanks.  I fixed the comments and added a test that unit-tests the nonpublic methods, including starting in the middle of a multibyte character.

dmk42 commented:
@ronawho - FYI, this will make significant changes to one performance graph, and slight changes to a few.

dmk42 commented:
Performance graph follow-up:
The results look good.

https://www.chapel-lang.org/perf/chapcs/?startdate=2019/05/01&enddate=2019/08/05&graphs=arrayofstringelementaccess,stringtemporarycopies,searchwithinastring

ronawho commented:
I'm happy with these improvements, but I did find myself wondering if we could further improve ascii-only strings by caching if we know it's ascii only. I was imagining something like:

```diff
diff --git a/modules/internal/String.chpl b/modules/internal/String.chpl
index 3cae2be509..b14e5f34b8 100644
--- a/modules/internal/String.chpl
+++ b/modules/internal/String.chpl
@@ -520,6 +520,7 @@ module String {
     // a locale object. Used when determining if we should make a remote transfer.
     var locale_id = chpl_nodeID; // : chpl_nodeID_t

+    var asciiOnly: int(8) = -1; // -1=unknown, 0=no, 1 = yes
     pragma "no doc"
     proc init() {
       // Let compiler insert defaults
@@ -707,6 +708,7 @@ module String {
                 string is correctly-encoded UTF-8.
       */
     proc numCodepoints {
+      if this.asciiOnly == 1 then {writeln("ascii only"); return this.length;}
       var localThis: string = this.localize();
       var n = 0;
       var i = 0;
@@ -716,6 +718,7 @@ module String {
           i += 1;
         n += 1;
       }
+      if i == n then this.asciiOnly = 1;
       return n;
     }


```

This is not an area I'm familiar with though, so I could be missing something that makes this obviously wrong.

dmk42 commented:
Good point.  I will bring that up at the deep dive.
